File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/f27128caccea0664241b6ec7efdcbeeb.go
package main

import (
	"fmt"
)

type Mint struct {
	value int
}

func inverse(a, m int) int {
	u, v := 0, 1
	for a != 0 {
		t := m / a
		m -= t * a
		a, m = m, a
		u -= t * v
		u, v = v, u
	}
	return u
}

type id0 struct {
	node []struct {
		w  int
		Wp Mint
		fp Mint
		tp Mint
	}
}

func (i *id0) unite(a, b struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
}) struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
} {
	res := struct {
		w  int
		Wp Mint
		fp Mint
		tp Mint
	}{}
	res.w = a.w + b.w
	res.Wp = a.Wp + b.Wp
	res.fp = a.fp + b.fp
	res.tp = a.tp + b.tp
	return res
}

func (i *id0) push(x, l, r int) {

}

func (i *id0) pull(x, z int) {
	i.node[x] = i.unite(i.node[x+1], i.node[z])
}

func (i *id0) build(x, l, r int) {
	if l == r {
		return
	}
	y := (l + r) >> 1
	z := x + ((y - l + 1) << 1)
	i.build(x+1, l, y)
	i.build(z, y+1, r)
	i.pull(x, z)
}

func (i *id0) buildWithValues(x, l, r int, v, a []int) {
	if l == r {
		i.node[x].apply(l, r, v[l], l, n-1-r, a[l])
		return
	}
	y := (l + r) >> 1
	z := x + ((y - l + 1) << 1)
	i.buildWithValues(x+1, l, y, v, a)
	i.buildWithValues(z, y+1, r, v, a)
	i.pull(x, z)
}

func (i *id0) get(x, l, r, ll, rr int) struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
} {
	if ll <= l && r <= rr {
		return i.node[x]
	}
	y := (l + r) >> 1
	z := x + ((y - l + 1) << 1)
	i.push(x, l, r)
	res := struct {
		w  int
		Wp Mint
		fp Mint
		tp Mint
	}{}
	if rr <= y {
		res = i.get(x+1, l, y, ll, rr)
	} else {
		if ll > y {
			res = i.get(z, y+1, r, ll, rr)
		} else {
			res = i.unite(i.get(x+1, l, y, ll, rr), i.get(z, y+1, r, ll, rr))
		}
	}
	i.pull(x, z)
	return res
}

func (i *id0) modify(x, l, r, ll, rr, v, id, rid, a int) {
	if ll <= l && r <= rr {
		i.node[x].apply(l, r, v, id, rid, a)
		return
	}
	y := (l + r) >> 1
	z := x + ((y - l + 1) << 1)
	i.push(x, l, r)
	if ll <= y {
		i.modify(x+1, l, y, ll, rr, v, id, rid, a)
	}
	if rr > y {
		i.modify(z, y+1, r, ll, rr, v, id, rid, a)
	}
	i.pull(x, z)
}

func (i *id0) findFirstKnowingly(x, l, r int, f func(struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
}) bool) int {
	if l == r {
		return l
	}
	i.push(x, l, r)
	y := (l + r) >> 1
	z := x + ((y - l + 1) << 1)
	var res int
	if f(i.node[x+1]) {
		res = i.findFirstKnowingly(x+1, l, y, f)
	} else {
		res = i.findFirstKnowingly(z, y+1, r, f)
	}
	i.pull(x, z)
	return res
}

func (i *id0) findFirst(x, l, r, ll, rr int, f func(struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
}) bool) int {
	if ll <= l && r <= rr {
		if !f(i.node[x]) {
			return -1
		}
		return i.findFirstKnowingly(x, l, r, f)
	}
	i.push(x, l, r)
	y := (l + r) >> 1
	z := x + ((y - l + 1) << 1)
	res := -1
	if ll <= y {
		res = i.findFirst(x+1, l, y, ll, rr, f)
	}
	if rr > y && res == -1 {
		res = i.findFirst(z, y+1, r, ll, rr, f)
	}
	i.pull(x, z)
	return res
}

func (i *id0) findLastKnowingly(x, l, r int, f func(struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
}) bool) int {
	if l == r {
		return l
	}
	i.push(x, l, r)
	y := (l + r) >> 1
	z := x + ((y - l + 1) << 1)
	var res int
	if f(i.node[z]) {
		res = i.findLastKnowingly(z, y+1, r, f)
	} else {
		res = i.findLastKnowingly(x+1, l, y, f)
	}
	i.pull(x, z)
	return res
}

func (i *id0) findLast(x, l, r, ll, rr int, f func(struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
}) bool) int {
	if ll <= l && r <= rr {
		if !f(i.node[x]) {
			return -1
		}
		return i.findLastKnowingly(x, l, r, f)
	}
	i.push(x, l, r)
	y := (l + r) >> 1
	z := x + ((y - l + 1) << 1)
	res := -1
	if rr > y {
		res = i.findLast(z, y+1, r, ll, rr, f)
	}
	if ll <= y && res == -1 {
		res = i.findLast(x+1, l, y, ll, rr, f)
	}
	i.pull(x, z)
	return res
}

func (i *id0) findFirst(ll, rr int, f func(struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
}) bool) int {
	return i.findFirst(0, 0, n-1, ll, rr, f)
}

func (i *id0) findLast(ll, rr int, f func(struct {
	w  int
	Wp Mint
	fp Mint
	tp Mint
}) bool) int {
	return i.findLast(0, 0, n-1, ll, rr, f)
}

func (i *id0) apply(l, r, v, id, rid, a int) {
	i.node[l].w = v
	i.node[l].Wp = Mint(v) * a
	i.node[l].fp = Mint(v) * id
	i.node[l].tp = Mint(v) * rid
}

func main() {
	var n, q int
	fmt.Scan(&n, &q)
	a := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i])
	}
	w := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&w[i])
	}
	st := id0{}
	st.node = make([]struct {
		w  int
		Wp Mint
		fp Mint
		tp Mint
	}, 2*n-1)
	st.buildWithValues(0, 0, n-1, w, a)
	for q > 0 {
		var x, y int
		fmt.Scan(&x, &y)
		if x < 0 {
			x = -x
			x--
			st.modify(x, x, y, x, n-1-x, a[x])
		} else {
			x--
			y--
			v := 0
			sum := st.get(x, y).w
			pos := st.findFirst(x, y, func(a struct {
				w  int
				Wp Mint
				fp Mint
				tp Mint
			}) bool {
				if (v+a.w)*2 >= sum {
					return true
				}
				v += a.w
				return false
			})
			l := st.get(x, pos)
			r := st.get(pos, y)
			res := Mint(0)
			res += r.Wp - r.fp - Mint(a[pos]-pos)*r.w
			res += Mint(a[pos]-pos)*l.w + l.fp - l.Wp
			fmt.Println(res)
		}
		q--
	}
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/f27128caccea0664241b6ec7efdcbeeb.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef long long ll;

typedef struct {
    ll w;
    ll Wp;
    ll fp;
    ll tp;
} node;

node unite(const node *a, const node *b) {
    node res;
    res.w = a->w + b->w;
    res.Wp = a->Wp + b->Wp;
    res.fp = a->fp + b->fp;
    res.tp = a->tp + b->tp;
    return res;
}

void push(node *tree, int x, int l, int r) {
    (void)tree;
    (void)x;
    (void)l;
    (void)r;
}

void pull(node *tree, int x, int z) {
    tree[x] = unite(&tree[x + 1], &tree[z]);
}

void build(node *tree, int x, int l, int r) {
    if (l == r) {
        return;
    }
    int y = (l + r) >> 1;
    int z = x + ((y - l + 1) << 1);
    build(tree, x + 1, l, y);
    build(tree, z, y + 1, r);
    pull(tree, x, z);
}

void build_with_values(node *tree, int x, int l, int r, const ll *v, const ll *a, int n) {
    if (l == r) {
        tree[x].w = v[l];
        tree[x].Wp = v[l] * a[l];
        tree[x].fp = v[l] * l;
        tree[x].tp = v[l] * (n - 1 - r);
        return;
    }
    int y = (l + r) >> 1;
    int z = x + ((y - l + 1) << 1);
    build_with_values(tree, x + 1, l, y, v, a, n);
    build_with_values(tree, z, y + 1, r, v, a, n);
    pull(tree, x, z);
}

node get(node *tree, int x, int l, int r, int ll, int rr) {
    if (ll <= l && r <= rr) {
        return tree[x];
    }
    int y = (l + r) >> 1;
    int z = x + ((y - l + 1) << 1);
    push(tree, x, l, r);
    node res;
    if (rr <= y) {
        res = get(tree, x + 1, l, y, ll, rr);
    } else {
        if (ll > y) {
            res = get(tree, z, y + 1, r, ll, rr);
        } else {
            res = unite(&get(tree, x + 1, l, y, ll, rr), &get(tree, z, y + 1, r, ll, rr));
        }
    }
    pull(tree, x, z);
    return res;
}

void modify(node *tree, int x, int l, int r, int ll, int rr, int v, int id, int rid, int a) {
    if (ll <= l && r <= rr) {
        tree[x].w = v;
        tree[x].Wp = v * a;
        tree[x].fp = v * id;
        tree[x].tp = v * rid;
        return;
    }
    int y = (l + r) >> 1;
    int z = x + ((y - l + 1) << 1);
    push(tree, x, l, r);
    if (ll <= y) {
        modify(tree, x + 1, l, y, ll, rr, v, id, rid, a);
    }
    if (rr > y) {
        modify(tree, z, y + 1, r, ll, rr, v, id, rid, a);
    }
    pull(tree, x, z);
}

int find_first_knowingly(node *tree, int x, int l, int r, bool (*f)(const node*)) {
    if (l == r) {
        return l;
    }
    push(tree, x, l, r);
    int y = (l + r) >> 1;
    int z = x + ((y - l + 1) << 1);
    int res;
    if (f(&tree[x + 1])) {
        res = find_first_knowingly(tree, x + 1, l, y, f);
    } else {
        res = find_first_knowingly(tree, z, y + 1, r, f);
    }
    pull(tree, x, z);
    return res;
}

int find_first(node *tree, int x, int l, int r, int ll, int rr, bool (*f)(const node*)) {
    if (ll <= l && r <= rr) {
        if (!f(&tree[x])) {
            return -1;
        }
        return find_first_knowingly(tree, x, l, r, f);
    }
    push(tree, x, l, r);
    int y = (l + r) >> 1;
    int z = x + ((y - l + 1) << 1);
    int res = -1;
    if (ll <= y) {
        res = find_first(tree, x + 1, l, y, ll, rr, f);
    }
    if (rr > y && res == -1) {
        res = find_first(tree, z, y + 1, r, ll, rr, f);
    }
    pull(tree, x, z);
    return res;
}

int find_last_knowingly(node *tree, int x, int l, int r, bool (*f)(const node*)) {
    if (l == r) {
        return l;
    }
    push(tree, x, l, r);
    int y = (l + r) >> 1;
    int z = x + ((y - l + 1) << 1);
    int res;
    if (f(&tree[z])) {
        res = find_last_knowingly(tree, z, y + 1, r, f);
    } else {
        res = find_last_knowingly(tree, x + 1, l, y, f);
    }
    pull(tree, x, z);
    return res;
}

int find_last(node *tree, int x, int l, int r, int ll, int rr, bool (*f)(const node*)) {
    if (ll <= l && r <= rr) {
        if (!f(&tree[x])) {
            return -1;
        }
        return find_last_knowingly(tree, x, l, r, f);
    }
    push(tree, x, l, r);
    int y = (l + r) >> 1;
    int z = x + ((y - l + 1) << 1);
    int res = -1;
    if (rr > y) {
        res = find_last(tree, z, y + 1, r, ll, rr, f);
    }
    if (ll <= y && res == -1) {
        res = find_last(tree, x + 1, l, y, ll, rr, f);
    }
    pull(tree, x, z);
    return res;
}

void id0(int n, int q) {
    ll *a = (ll*)malloc(n * sizeof(ll));
    ll *w = (ll*)malloc(n * sizeof(ll));
    for (int i = 0; i < n; i++) {
        scanf("%lld", &a[i]);
    }
    for (int i = 0; i < n; i++) {
        scanf("%lld", &w[i]);
    }
    node *st = (node*)malloc((2 * n - 1) * sizeof(node));
    build_with_values(st, 0, 0, n - 1, w, a, n);
    while (q--) {
        int x, y;
        scanf("%d%d", &x, &y);
        if (x < 0) {
            x = -x;
            x--;
            modify(st, 0, 0, n - 1, x, x, y, x, n - 1 - x, a[x]);
        } else {
            x--; y--;
            ll v = 0;
            ll sum = get(st, 0, 0, n - 1, x, y).w;
            int pos = find_first(x, y, st, 0, 0, n - 1, [&](const node *a) {
                if ((v + a->w) * 2 >= sum) return true;
                v += a->w;
                return false;
            });
            if (pos != -1) {
                node l = get(st, 0, 0, n - 1, x, pos);
                node r = get(st, 0, 0, n - 1, pos, y);
                ll res = 0;
                res += r.Wp - r.fp - (a[pos] - pos) * r.w;
                res += (a[pos] - pos) * l.w + l.fp - l.Wp;
                printf("%lld\n", res);
            }
        }
    }
    free(a);
    free(w);
    free(st);
}

int main() {
    int n, q;
    scanf("%d%d", &n, &q);
    id0(n, q);
    return 0;
}
